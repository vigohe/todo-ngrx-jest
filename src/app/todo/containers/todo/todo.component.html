<app-input (add)="addTodo($event)"></app-input>
<app-list [todos]="todos | async" (toggle)="onToggle($event)" (remove)="onRemove($event)"></app-list>
