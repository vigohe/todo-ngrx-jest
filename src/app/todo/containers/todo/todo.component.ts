import { Component, OnInit } from '@angular/core';
import { Store } from '@ngrx/store';
import * as fromStore from '../../reducers/todo.reducer';
import { AddTodo, ToggleTodo, RemoveTodo } from '../../actions/todo.actions';
import { Todo } from '../../models/todo';
import { Observable } from 'rxjs';
import { State } from '../../../reducers';

@Component({
  selector: 'app-todo',
  templateUrl: './todo.component.html',
  styleUrls: ['./todo.component.css']
})
export class TodoComponent implements OnInit {
  public todos: Observable<Todo[]>;

  constructor(private store: Store<State>) { }

  ngOnInit() {
    this.todos = this.store.select((state: State) => state.todo.entities );
  }

  addTodo(newTodo: string) {
    this.store.dispatch(new AddTodo(newTodo));
  }

  onToggle(todo: Todo) {
    this.store.dispatch(new ToggleTodo(todo));
  }

  onRemove(todo: Todo) {
    this.store.dispatch(new RemoveTodo(todo));
  }
}
