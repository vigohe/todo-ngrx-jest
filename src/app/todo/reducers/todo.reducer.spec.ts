import { reducer, initialState } from "./todo.reducer";
import {
  TodoActionTypes,
  AddTodo,
  RemoveTodo,
  ToggleTodo
} from "../actions/todo.actions";

describe("Todo Reducer", () => {

  const state = {
    entities: [
      {
        id: 1,
        text: "READ A BOOK",
        complete: false
      },
      {
        id: 2,
        text: "WATCH A MOVIE",
        complete: false
      }
    ]
  };

  describe("unknown action", () => {
    it("should return the initial state", () => {
      const action = {} as any;

      const result = reducer(undefined, action);

      expect(result).toBe(initialState);
    });
  });

  describe("AddTodo action", () => {
    it("should return a new Todo", () => {
      const action = new AddTodo("NEW TODO");
      const result = reducer(initialState, action);
      expect(result.entities).toHaveLength(1);
      expect(result.entities[0].text).toBe("NEW TODO");
    });
  });

  describe("RemoveTodo action", () => {
    it("should remove a Todo", () => {
      const action = new RemoveTodo(state.entities[0]);

      const result = reducer(state, action);
      expect(result.entities).toHaveLength(1);
      expect(result.entities).toMatchSnapshot();
    });
  });

  describe("ToggleTodo action", () => {
    it("should toggle a Todo", () => {
      const action = new ToggleTodo(state.entities[1]);
      const result = reducer(state, action);
      expect(result.entities[1].complete).toBe(true);
      expect(result.entities).toMatchSnapshot();
    });
  });
});
