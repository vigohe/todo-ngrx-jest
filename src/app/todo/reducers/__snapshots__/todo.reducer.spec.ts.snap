// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo Reducer RemoveTodo action should remove a Todo 1`] = `
Array [
  Object {
    "complete": false,
    "id": 2,
    "text": "WATCH A MOVIE",
  },
]
`;

exports[`Todo Reducer ToggleTodo action should toggle a Todo 1`] = `
Array [
  Object {
    "complete": false,
    "id": 1,
    "text": "READ A BOOK",
  },
  Object {
    "complete": true,
    "id": 2,
    "text": "WATCH A MOVIE",
  },
]
`;
